[sn32] ct16: further chibios integration

dexter93 opened the pull request:
ChibiOS integration of CT[16] in general purpose timer and pwm driver. The driver supports both CT16B0 and CT16B1

dexter93 committed:
[sn32] ct16: further chibios integration

general purpose and pwm driver
also introduce a reset function

dexter93 committed:
build ct16 driver

dexter93 committed:
[sn32] have some isr handling

dexter93 committed:
240: config the board for pwm

dexter93 committed:
Revert "240: config the board for pwm"
probably best to do in pwm driver
This reverts commit c09059a8ba60ea1832ea14f4d557bd8e22df3fd7.

dexter93 committed:
ct: logic fix. remove unecessary include

dexter93 committed:
ct: typos + use notifications

dexter93 committed:
ct: the periodic notification should not stop the timer

dexter93 commented:
pwm driver init example
```/* PWM configuration structure. We use timer CT16B1 with 24 channels. */
static const PWMConfig pwmcfg = {
    3000000,        /* 3 MHz PWM clock frequency. */
    256,            /* PWM period is 256 cycles. */
    0,
    {
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0},
        {PWM_OUTPUT_ACTIVE_HIGH, 0}
    },
    0
};
```

pwm driver start example 
```  
 pwmStart(&PWMD1, &pwmcfg);
```
further pwm driver refs [here](http://chibios.sourceforge.net/docs3/hal/group___p_w_m.html)

gpt driver refs [here](http://chibios.sourceforge.net/docs3/hal/group___g_p_t.html)

sn32 flags for halconf
```
HAL_USE_GPT
SN32_GPT_USE_CT16B0
SN32_GPT_USE_CT16B1

HAL_USE_PWM
SN32_PWM_USE_CT16B1
```

board conf
```
SN32_HAS_CT16B0
SN32_HAS_CT16B1
```

dexter93 committed:
ct: pfpa assignment

dexter93 committed:
ct: pwm:  enable channel only if configured as active

dexter93 committed:
ct: pwm: periodic tic should track the timer

dexter93 committed:
derp

dexter93 committed:
ct: our chip supports IRQ priority level [0-3]

dexter93 committed:
isr: use the appropriate priority for each device

dexter93 committed:
ct: pwm: actually handle the channels

dexter93 committed:
ct: we define the # of channels already so..

dexter93 committed:
isr: we handle this on the specific drivers
disable for now

dexter93 committed:
ct: pwm: don't override TC. Call a counter reset

IsaacDynamo commented on os/hal/ports/SN32/LLD/SN32F24xB/CT/CT16B0.c:
```diff
@@ -57,7 +57,22 @@ void CT16B0_Init (void)
+* Return		: None
+* Note			: None
+*****************************************************************************/
+void CT16B0_Reset (void)
```
Reset could also imply resetting the peripheral, consider renaming to ResetCounter or something like that.

dexter93 replied:
Done with https://github.com/SonixQMK/ChibiOS-Contrib/pull/30/commits/8499f50478da6fde760bea5d05c240bbaf3aaf36

IsaacDynamo commented on os/hal/ports/SN32/LLD/SN32F24xB/CT/hal_pwm_lld.c:
```diff
@@ -0,0 +1,997 @@
+  switch(channel){
+    case 0:
+      pwmp->ct->IC |= mskCT16_MR0IC;
+      pwmp->ct->PWMIOENB |= mskCT16_PWM0IOEN_DIS;
```
This will not work, ORing with 0 (mskCT16_PWM0IOEN_DIS is 0<<0) is a NOP, and will not clear the bit.
I think you mean something like `pwmp->ct->PWMIOENB &= ~mskCT16_PWM0IOEN_EN`

dexter93 replied:
nice find. fixed https://github.com/SonixQMK/ChibiOS-Contrib/pull/30/commits/b3c911f27a531e31915b1f263ed0f8396b134f88

IsaacDynamo commented on os/hal/ports/SN32/LLD/SN32F24xB/CT/hal_pwm_lld.c:
```diff
@@ -0,0 +1,997 @@
+                                          pwmchannel_t channel) {
+  switch(channel){
+    case 0:
+      pwmp->ct->MCTRL |= mskCT16_MR0IE_DIS;
```
This is also an OR with 0, see previous comment

IsaacDynamo commented on os/hal/ports/SN32/SN32F240B/sn32_isr.c:
```diff
@@ -0,0 +1,126 @@
+ *
+ * @notapi
+ */
+void irqInit(void) {
```
What is the usecase of irqInit() and irqDeinit(), and the files sn32_isr.c / .h?

Shoudn't nvicEnableVector() and nvicDisableVector() be called by the gpt and ct drivers?

What are the SN32_CT16B0_SUPPRESS_ISR defines?

dexter93 replied:
These were leftovers from the awfully generic stm32 hal I traced for this. I am still tracing for a use case scenario, but it looks like I will probably amend and remove the isr.c/h completely

dexter93 committed:
ct: pwm: Set the prescaler properly

PRE is the max value of PC. PC increments on every tick.
When PC reaches PRE, TC increments and the timer overflows
Periodic timer is software controlled

dexter93 committed:
ct: gpt: PRE is the interval. Call a counter reset

dexter93 committed:
ct: Rename reset funtion to be more precise

dexter93 committed:
ct: pwm: Invert the channel disable logic

dexter93 committed:
ct: pwm: autoreload on period match

dexter93 committed:
ct: pwm: we only care for the last 25bits of IC

dexter93 committed:
ct: gpt: we only care for the last 25bits of IC

dexter93 committed:
ct: pwm: init config as hw defaults

dexter93 committed:
ct: pwm: mr24 is the driver callback

dexter93 committed:
Revert "isr: we handle this on the specific drivers"

This reverts commit 9ca061f170d9523d0e4e42011881f3d0b8dd1599.

dexter93 committed:
Revert "[sn32] have some isr handling"

This reverts commit cf45020072ea828522f2de2f6d5d2521398a61cd.

dexter93 committed:
ct: pwm: update the periodic notification

dexter93 committed:
ct: gpt: use MR0

dexter93 committed:
ct: priority bump

dexter93 committed:
Revert "ct: priority bump"
3 is probably fine
This reverts commit d2a861097a43f367ffbd15f6fc5d6747d9f14426.

dexter93 committed:
sn32: pwm: introduce oneshot mode

enable it with by defining SN32_PWM_USE_ONESHOT TRUE

dexter93 committed:
pwm: reset: we have a function for that

dexter93 committed:
explicitly enable pwm channels

dexter93 committed:
bugfix for bad PWM_OUTPUT_MASK

if it is DISABLED, bad things occur
Only ever identify and use PWM_OUTPUT_ACTIVE_HIGH and PWM_OUTPUT_ACTIVE_LOW

dexter93 committed:
periodic notification: clear and disable it

dexter93 committed:
we definitely need this to only run at first init

dexter93 committed:
ct: pwm: support 23 channel chips

dexter93 commented:
This is almost ready to go.

When https://github.com/SonixQMK/ChibiOS-Contrib/pull/35 is merged, a chip check will be done inside `sn32_ct.h ` for channel selection, and in pwm hal to disable PFPA access for 260 chips.

dexter93 committed:
ct: support 240b and 260 chips

dexter93 committed:
make the macro check build

stdvar commented:
1. update indents in pwm_lld_start (see comment above)
2. for majority of cases callbacks are not used per channel (for now at least). Implement optimization to check 1-23 altogether and if any bit is raised then check individually one by one. that will save on instructions for every interrupt.
3. mskCT16_PWM2MODE_1 should be called PWM_OUTPUT_ACTIVE_LOW (see datasheet)

dexter93 committed:
pwm: correctly set the logic level

dexter93 committed:
pwm: speed improvements
in our shared matrix driver, we only care about the last callback.
having a shortcut to that if no other flag is raised will improve speed

+ident and comment

dexter93 committed:
Revert "pwm: speed improvements"

This reverts commit e143544b807dd860d26548c803503521450822b8.
